ENT-4064: Added diagnose command to cf-check

olehermanse opened the pull request:
Following up from https://github.com/cfengine/core/pull/3367

This command assesses the health of embedded (LMDB) databases.
It forks a subprocess for each listed database,	iterates through
all data in the	database (printing it to /dev/null) and	returns
non-zero exit code if any database is corrupt.

olehermanse commented:
@cf-bottom let's build this in jenkins with exotics.

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1660)](https://ci.cfengine.com//job/pr-pipeline/1660/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1660/

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1663)](https://ci.cfengine.com//job/pr-pipeline/1663/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1663/

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1665)](https://ci.cfengine.com//job/pr-pipeline/1665/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1665/

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1668)](https://ci.cfengine.com//job/pr-pipeline/1668/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1668/

vpodzime commented on cf-check/diagnose.c:
```diff
@@ -0,0 +1,353 @@
+#include <lmdb.h>
+#include <sys/wait.h>
+#include <signal.h>
+#include <lmdump.h>
```
Duplicate include of `lmdump.h`.

olehermanse replied:
Fixed, thanks!

vpodzime commented on cf-check/diagnose.c:
```diff
@@ -0,0 +1,353 @@
+// static bool is_recognized_errno(int r)
+// {
+//     return (strerror_or_null(r) != NULL);
+// }
```
This can be dropped now, I guess.

olehermanse replied:
Yes, thanks.

vpodzime commented on cf-check/diagnose.c:
```diff
@@ -0,0 +1,353 @@
+    }
+
+    const char *unknown = "Unknown error";
+    if (strncmp(strerror_string, unknown, strlen(unknown)) == 0)
```
Not important here, but you could optimize this by doing:

```
const char unknown[] = "Unknown error";
if (strncmp(strerror_string, unknown, sizeof(unknown) - 1) == 0)
```

olehermanse replied:
Yes, that's neat.

vpodzime commented on cf-check/diagnose.c:
```diff
@@ -0,0 +1,353 @@
+    case SIGABRT:
+        return CF_CHECK_SIGNAL_ABORT;
+    // case SIGEMT:
+    //     return CF_CHECK_SIGNAL_EMULATE_INSTRUCTION;
```
Please remove this or add a comment about where it appeared.

olehermanse replied:
Added a comment

vpodzime commented on cf-check/diagnose.c:
```diff
@@ -0,0 +1,353 @@
+    case SIGWINCH:
+        return CF_CHECK_SIGNAL_WINDOW_SIZE_CHANGE;
+    // case SIGINFO:
+    //     return CF_CHECK_SIGNAL_STATUS_REQUEST;
```
Same here.

olehermanse replied:
Added a comment

vpodzime commented on cf-check/diagnose.c:
```diff
@@ -0,0 +1,353 @@
+    case MDB_BAD_VALSIZE:
+        return CF_CHECK_LMDB_WRONG_KEY_OR_VALUE_SIZE;
+    // case MDB_BAD_DBI:
+    //     return CF_CHECK_LMDB_BAD_DBI;
```
Same here.

olehermanse replied:
Added a comment

vpodzime commented on cf-check/lmdump.c:
```diff
@@ -7,6 +7,7 @@
 #include <time.h>
 #include <string.h>
 #include <lmdb.h>
+#include <errno.h>
```
Why is `errno.h` suddenly needed here?

olehermanse replied:
It's not, removed.

vpodzime commented on cf-check/lmdump.c:
```diff
@@ -92,7 +93,7 @@ int lmdump(lmdump_mode mode, const char *file)
     if (rc) return lmdump_report_error(rc);
 
-    rc = mdb_env_open(env, file, MDB_NOSUBDIR, 0644);
+    rc = mdb_env_open(env, file, MDB_NOSUBDIR | MDB_RDONLY, 0644);
```
This looks a little unrelated to the commit.

olehermanse replied:
A necessary change, but maybe I should add a better message or add it as a separate commit.

olehermanse replied:
Put in a separate commit.

vpodzime commented on travis-scripts/valgrind.sh:
```diff
@@ -165,8 +165,9 @@ echo "Checking that serverd and execd PIDs are still correct/alive:"
-echo "Killing valgrind"
+echo "Killing valgrind cf-execd"
 kill $exec_pid
+echo "Killing valgrind cf-serverd"
```
Also unrelated.

olehermanse replied:
Changed commits to better reflect this change.

olehermanse commented:
@cf-bottom let's test this one more time in jenkins ;)

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=1675)](https://ci.cfengine.com//job/pr-pipeline/1675/)

https://ci.cfengine.com/job/pr-pipeline/1675/

hyc commented:
Out of curiosity, why didn't you just use mdb_dump instead of writing lmdump?

olehermanse commented:
@hyc I didn't write `lmdump`, just moved it from another repo. I don't know why it was written in the first place, but it is several years old. Perhaps because we don't have lmdb-utils as a dependency, and writing something small and simple is usually easier than introducing a new dependency. But this is just speculation from my side. I added it as a command here for backwards compatibility, and so I can remove `lmdump` patch from our buildscripts.
